### PR TITLE
api, pdctl: add HTTP Component signature

### LINKED ISSUE #4490: Add HTTP Component Signature
## Development Task
refer to #4480

### PR DESCRIPTION
**Pull request by CabinfeverB**
### What problem does this PR solve?
close #4490  and related to #3111

This PR is going to support HTTP component signature

### What is changed and how it works?
It adds a `RoundTripper` for adding a component signature header to every request and applies it to pdctl. (other components outside PD could add header by themselves)

For the test, this PR creates a test server and registers a temp handler for `/pd/api/v1/health`(just for test) and sends the request by health pdctl cmd.

### Check List

Tests

- Unit test

### Release note

```release-note
add HTTP Component signature
```

### CODE REVIEW

File: pkg/apiutil/apiutil.go [hunk 1]
```diff
@@ -27,6 +27,14 @@ import (
 	"github.com/unrolled/render"
 )
 
+var (
+	// ComponentSignatureKey is used for http request header key
+	// to identify component signature
+	ComponentSignatureKey = "component"
+	// ComponentAnonymousValue identify anonymous request source
```
Comment (disksing) [inline pkg/apiutil/apiutil.go:34]:
```suggestion
	// ComponentAnonymousValue identifies anonymous request source
```

File: pkg/apiutil/apiutil.go [hunk 2]
```diff
@@ -127,3 +135,26 @@ func ErrorResp(rd *render.Render, w http.ResponseWriter, err error) {
 		rd.JSON(w, http.StatusInternalServerError, err.Error())
 	}
 }
+
+// GetComponentNameOnHTTP return component name from Request Header
```
Comment (disksing) [inline pkg/apiutil/apiutil.go:139]:
```suggestion
// GetComponentNameOnHTTP returns component name from Request Header
```

**Review by disksing — APPROVED** [review]

File: pkg/apiutil/apiutil.go [hunk 3]
```diff
@@ -27,6 +27,14 @@ import (
 	"github.com/unrolled/render"
 )
 
+var (
+	// ComponentSignatureKey is used for http request header key
+	// to identify component signature
+	ComponentSignatureKey = "component"
```
Comment (rleungx) [inline pkg/apiutil/apiutil.go:33]: no need to export?

File: tools/pd-ctl/pdctl/command/global.go [hunk 4]
```diff
@@ -25,11 +25,17 @@ import (
...
 	"github.com/spf13/cobra"
+	"github.com/tikv/pd/pkg/apiutil"
 	"go.etcd.io/etcd/pkg/transport"
 )
 
 var (
-	dialClient = &http.Client{}
+	dialClient = &http.Client{
+		Transport: &apiutil.ComponentSignatureRoundTripper{
+			Component: "pdctl",
```
Comment (rleungx) [inline tools/pd-ctl/pdctl/command/global.go:35]: How about making pdctl a constant?

File: pkg/apiutil/apiutil.go [hunk 5]
```diff
@@ -127,3 +135,26 @@ func ErrorResp(rd *render.Render, w http.ResponseWriter, err error) {
...
+	componentName := r.Header.Get(ComponentSignatureKey)
+	if componentName == "" {
+		componentName = ComponentAnonymousValue
+	}
+	return componentName
+}
+
+// ComponentSignatureRoundTripper is used to add component signature in HTTP header
+type ComponentSignatureRoundTripper struct {
+	Proxied   http.RoundTripper
```
Comment (bufferflies) [inline pkg/apiutil/apiutil.go:150]: this properties may be private and add construct function to set them.

**Review by nolouch — COMMENTED** [review]
lgtm. but ci failed.

File: pkg/apiutil/apiutil.go [hunk 6]
```diff
@@ -127,3 +135,34 @@ func ErrorResp(rd *render.Render, w http.ResponseWriter, err error) {
 		rd.JSON(w, http.StatusInternalServerError, err.Error())
 	}
 }
+
+// GetComponentNameOnHTTP returns component name from Request Header
+func GetComponentNameOnHTTP(r *http.Request) string {
+	componentName := r.Header.Get(componentSignatureKey)
+	if componentName == "" {
```
Comment (bufferflies) [inline pkg/apiutil/apiutil.go:142]:
```suggestion
	if len(componentName)==0 {
```

File: pkg/apiutil/apiutil.go [hunk 7]
```diff
@@ -127,3 +135,34 @@ func ErrorResp(rd *render.Render, w http.ResponseWriter, err error) {
...
+		componentName = componentAnonymousValue
+	}
+	return componentName
+}
+
+// ComponentSignatureRoundTripper is used to add component signature in HTTP header
+type ComponentSignatureRoundTripper struct {
+	proxied   http.RoundTripper
+	component *string
```
Comment (bufferflies) [inline pkg/apiutil/apiutil.go:151]: why not use point as receiver?

**Review by bufferflies — APPROVED** [review]

**Pull request merged**